### PR TITLE
Mixup docment

### PR DESCRIPTION
**Pull request by kevinbird15**
I think my local branch is slightly ahead of the docment branch because of recent fastai changes  Because of that, I'm not going to worry too much about the files that changed besides the three mixup unless somebody wants to help me fix it so it only changes the files I actually am meaning to:

### CODE REVIEW

**Comment by warner-benjamin** [comment]
A couple of comments:

`from __future__ import annotations` shouldn't be added (yet).

I think this would be better formatted as:
```python
class MixUp(MixHandler):
    "Implementation of https://arxiv.org/abs/1710.09412"
    def __init__(self,
        alpha=.4 # Determine `Beta` Distribution (0.,inf]
    ): 
        super().__init__(alpha)
```
and the same for similarly formatted methods.

And for `rand_bbox` I think if there are newlines there should be docments for those args.

**Comment by kevinbird15** [comment]
> This references issue: #49
this should actually be #48 I believe

**Comment by hamelsmu** [comment]
Merging this could cause conflicts for those PRs you merge after this one - the diff on this PR is extremely large and is likely to cause lots of confusion for subsequent PRs

@kevinbird15 I would fork Zach's repo instead and replay your changes (looks like only this one notebook) on top of that

**Comment by kevinbird15** [comment]
Yeah, my git skills kind of failed me on this... I couldn't figure out how to discard those changes that were in the fastai master branch and keep the changes that I made

File: fastai/callback/mixup.py [hunk 1]
```diff
@@ -1,21 +1,33 @@
 # AUTOGENERATED! DO NOT EDIT! File to edit: nbs/19_callback.mixup.ipynb (unless otherwise specified).
 
+
+from __future__ import annotations
+
+
```
Comment (muellerzr) [inline fastai/callback/mixup.py:6]:
```suggestion
```

File: fastai/callback/mixup.py [hunk 2]
```diff
@@ -1,21 +1,33 @@
...
 from torch.distributions.beta import Beta
 
 # Cell
-def reduce_loss(loss, reduction='mean'):
+#nbdev_comment from __future__ import annotations
+
+# Cell
+def reduce_loss(
+    loss:Tensor,
+    reduction:(str|None)='mean' # Loss reduction type
```
Comment (muellerzr) [inline fastai/callback/mixup.py:19]:
```suggestion
    reduction:(str, None)='mean' # Loss reduction type
```

File: fastai/callback/mixup.py [hunk 3]
```diff
@@ -1,21 +1,33 @@
...
     "Reduce the loss based on `reduction`"
     return loss.mean() if reduction == 'mean' else loss.sum() if reduction == 'sum' else loss
 
 # Cell
 class MixHandler(Callback):
     "A handler class for implementing `MixUp` style scheduling"
     run_valid = False
-    def __init__(self, alpha=0.5):
+    def __init__(self,
+                 alpha:float=0.5 # Determine `Beta` Distribution (0.,inf]
```
Comment (muellerzr) [inline fastai/callback/mixup.py:29]:
```suggestion
                 alpha:float=0.5 # Determine `Beta` Distribution in range (0.,inf]
```

File: fastai/callback/mixup.py [hunk 4]
```diff
@@ -40,7 +52,9 @@ def lf(self, pred, *yb):
 # Cell
 class MixUp(MixHandler):
     "Implementation of https://arxiv.org/abs/1710.09412"
-    def __init__(self, alpha=.4): super().__init__(alpha)
+    def __init__(self,
+                 alpha=.4 # Determine `Beta` Distribution (0.,inf]
```
Comment (muellerzr) [inline fastai/callback/mixup.py:56]: Same note as earlier

File: fastai/callback/mixup.py [hunk 5]
```diff
@@ -40,7 +52,9 @@ def lf(self, pred, *yb):
...
+                ): super().__init__(alpha)
```
Comment (muellerzr) [inline fastai/callback/mixup.py:57]:
```suggestion
                ): 
            super().__init__(alpha)
```

File: fastai/callback/mixup.py [hunk 6]
```diff
@@ -57,7 +71,9 @@ def before_batch(self):
 # Cell
 class CutMix(MixHandler):
     "Implementation of https://arxiv.org/abs/1905.04899"
-    def __init__(self, alpha=1.): super().__init__(alpha)
+    def __init__(self,
+                 alpha:float=1. # Determine `Beta` Distribution (0.,inf]
+                ): super().__init__(alpha)
```
Comment (muellerzr) [inline fastai/callback/mixup.py:76]: Same notes as earlier

File: fastai/callback/mixup.py [hunk 7]
```diff
@@ -70,7 +86,11 @@ def before_batch(self):
             ny_dims = len(self.y.size())
             self.learn.yb = tuple(L(self.yb1,self.yb).map_zip(torch.lerp,weight=unsqueeze(self.lam, n=ny_dims-1)))
 
-    def rand_bbox(self, W, H, lam):
+    def rand_bbox(self,
+                  W,
+                  H,
+                  lam # lambda sample from Beta Distribution
```
Comment (muellerzr) [inline fastai/callback/mixup.py:92]: Could we add types to what `W` and `H` are? Or in a docment

**Review by muellerzr — CHANGES_REQUESTED** [review]
Made a few notes

File: fastai/callback/mixup.py [hunk 8]
```diff
@@ -7,15 +7,20 @@
 from torch.distributions.beta import Beta
 
 # Cell
-def reduce_loss(loss, reduction='mean'):
+def reduce_loss(
+    loss:Tensor,
+    reduction:(str, None)='mean' # Loss reduction type
```
Comment (kevinbird15) [inline fastai/callback/mixup.py:12]: Would it be good to list all options here since there are only three?

Comment (warner-benjamin) [reply fastai/callback/mixup.py:12]: Or could state:
```suggestion
    reduction:str='mean' # PyTorch loss reduction
```
Matching PyTorch, the type should only be str

Comment (muellerzr) [reply fastai/callback/mixup.py:12]: Agree with @warner-benjamin on this one.

File: fastai/callback/mixup.py [hunk 9]
```diff
@@ -31,7 +36,11 @@ def after_cancel_train(self):
     def after_cancel_fit(self):
         self.after_train()
 
-    def lf(self, pred, *yb):
+    def lf(self,
+        pred,
+        *yb
+    ):
```
Comment (muellerzr) [inline fastai/callback/mixup.py:42]: If there are no docments, keep on the same line

File: fastai/callback/mixup.py [hunk 10]
```diff
@@ -31,7 +36,11 @@ def after_cancel_train(self):
...
+        "lf is a loss function"
```
Comment (muellerzr) [inline fastai/callback/mixup.py:43]: Can we be a bit more accurate in what this is? While it is a loss function, it's a bit more than that

**Review by muellerzr — APPROVED** [review]
Make those few nits and then I approve @kevinbird15!

**Comment by kevinbird15** [comment]
Alright, I think all of the changes are done.  Just need @warner-benjamin's approval (or disapproval)

**Comment by kevinbird15** [comment]
> There are missing doc strings in methods, including methods like `Mixup.before_batch` which are not simply calling other methods.

I have added docstrings to everything, let me know if you want anything changed

> Also, could you remove the warnings/errors committed in the mixup notebook?

I don't know how to do this without modifying code.  I did create an issue [here](https://github.com/fastai/fastai/issues/3588) since fixing the issue will require a code change, I don't think I should do that in this repo.  Is there another way to fix this?

**Review by warner-benjamin — APPROVED** [review]
Looks good. Thanks!
